### ISSUE TITLE
Typed arrays

### ISSUE THREAD

**Issue by linlexing** [robertkrimen/otto#89]
When can support Typed arrays?

**Comment by stevenh** [comment]
Do you have an example of what your trying to do?

**Comment by linlexing** [comment]
"Typed Array" is good and consume less memory than the "array " performance. "Typed Array" for reading and writing the file contents , as well as for data encryption and decryption.
```javascript
// From a length
var uint8 = new Uint8Array(2);
uint8[0] = 42;
console.log(uint8[0]); // 42
console.log(uint8.length); // 2
console.log(uint8.BYTES_PER_ELEMENT); // 1

// From an array
var arr = new Uint8Array([21,31]);
console.log(arr[1]); // 31

// From another TypedArray
var x = new Uint8Array([21, 31]);
var y = new Uint8Array(x);
console.log(y[0]); // 21

// From an ArrayBuffer
var buffer = new ArrayBuffer(8);
var z = new Uint8Array(buffer, 1, 4);
```

**Comment by stevenh** [comment]
Apart from the BYTES_PER_ELEMENT, that should just work.

For the passing arrays in you'll need: https://github.com/robertkrimen/otto/pull/108

**Comment by boomshroom** [comment]
`undefined ReferenceError: Float32Array is not defined`

**Comment by ble** [comment]
References:
MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Typed_arrays
Relevant section from Draft ECMA-262 6th Edition (April 3, 2015): http://people.mozilla.org/~jorendorff/es6-draft.html#sec-typedarray-objects

**Comment by robertkrimen** [comment]
I thought about this briefly before. It would be nice to have with Go's natives types, but I don't know of a straightforward way (leveraging Go) to present "typed views" of the same data.

**Comment by ble** [comment]
There is a standard library way of reading arbitrary bytes as a sequence of fixed-size values: http://golang.org/pkg/encoding/binary/#Read

It seems designed for bulk transform, not for providing a typed view of bytes.  It might not perform very well if it gets called repeatedly, as you'd want to do for providing this interface.  Docs for the package say "This package favors simplicity over efficiency."

There's always the ````package unsafe```` way, too.

**Comment by 3dseals** [comment]
I got the error:
'Uint8Array' is not defined
'ArrayBuffer' is not defined
when require [github.com/dcodeIO/ProtoBuf.js ](https://github.com/dcodeIO/ProtoBuf.js)
I thought slices can be converted toArrayBuffer
Please refer to:
| Go type               | JavaScript type       | Conversions back to interface{} |
| --------------------- | --------------------- | ------------------------------- |
| bool                  | Boolean               | bool                            |
| integers and floats   | Number                | float64                         |
| string                | String                | string                          |
| []int8                | Int8Array             | []int8                          |
| []int16               | Int16Array            | []int16                         |
| []int32, []int        | Int32Array            | []int                           |
| []uint8               | Uint8Array            | []uint8                         |
| []uint16              | Uint16Array           | []uint16                        |
| []uint32, []uint      | Uint32Array           | []uint                          |
| []float32             | Float32Array          | []float32                       |
| []float64             | Float64Array          | []float64                       |
| all other slices      | Array                 | []interface{}                   |
| arrays                | see slice type        | see slice type                  |
| functions             | Function              | func(...interface{}) *js.Object |
| time.Time             | Date                  | time.Time                       |
| -                     | instanceof Node       | *js.Object                      |
| maps, structs         | instanceof Object     | map[string]interface{}          |
form [gopherjs](https://godoc.org/github.com/gopherjs/gopherjs/js)

by the way What time can use the ArrayBuffer

**Comment by boomshroom** [comment]
A little off topic, but Int32Array always gets converted to a []int? Even on a 64 bit machine where int means int64?

**Comment by hochhaus** [comment]
@stevenh Does your comment mean that typed arrays are currently implemented in otto?

When I try to to do something simple with types arrays (eg: `var uint8 = new Uint8Array(2);`) I receive a `ReferenceError: 'Uint8Array' is not defined` error.

**Comment by hochhaus** [comment]
Looking at the code a bit I'm reasonably confident that no attempt has been made to implement typed arrays. Could this limitation be documented in the [caveat-emptor](https://github.com/robertkrimen/otto#caveat-emptor) section of the README?

**Comment by deoxxa** [comment]
Types Arrays aren't part of ES5, and I don't think we need to start listing all the ES6 features otto doesn't support.

**Comment by hochhaus** [comment]
That is fair, how about listing that otto targets ES5 then?

Typed arrays are relatively ubiquitous at this point (much more so than most of ES6) so this seems like it might be a common pitfall.

**Comment by deoxxa** [comment]
Yep, fair call. I actually thought it was listed clearly already, but I appear to be mistaken! I'd happily merge a change to make this more obvious in the readme and the inline docs - could you put that in for me, @hochhaus?

**Comment by hochhaus** [comment]
Thanks @deoxxa. Change https://github.com/robertkrimen/otto/pull/216 sent.

**Comment by stevenh** [comment]
Typed arrays are part of ES6 so not currently supported, but it should be possible to do.
